src/Raven.Server/Documents/PeriodicBackup/GoogleCloudStorage/RavenGoogleCloudStorageClient.cs

```diff
@@ -0,0 +1,158 @@
+
+namespace Raven.Server.Documents.PeriodicBackup.GoogleCloudStorage
+{
+    public class RavenGoogleCloudStorageClient : IDisposable
+    {
+        private readonly StorageClient _client;
+        private readonly string _projectId;
+        protected readonly CancellationToken CancellationToken;
+        private readonly string _bucketName;
+        private readonly Progress _progress;
+
+        private const string ProjectIdPropertyName = "project_id";
+
+        public RavenGoogleCloudStorageClient(string jsonCredential, string bucketName, CancellationToken? cancellationToken = null, Progress progress = null)
+        {
+            try
+            {
+                _client = StorageClient.Create(GoogleCredential.FromJson(jsonCredential));
+            }
+            catch (Exception e)
+            {
+                throw new ArgumentException("Wrong format for account key.", e);
+            }
+
+            if (JObject.Parse(jsonCredential).TryGetValue(ProjectIdPropertyName, StringComparison.InvariantCultureIgnoreCase, out var value))
+            {
+                _projectId = value.Value<string>();
+            }
+
+            _bucketName = bucketName;
+            CancellationToken = cancellationToken ?? CancellationToken.None;
+
+            _progress = progress;
+        }
+
+        public async Task<Object> UploadObjectAsync(string fileName, Stream stream, Dictionary<string, string> metadata = null)
+        {
+
+            return await _client.UploadObjectAsync(
+                new Object
+                {
+                    Bucket = _bucketName,
+                    Name = fileName,
+                    ContentType = "application/octet-stream",
+                    Metadata = metadata
+                }, stream,
+                cancellationToken: CancellationToken,
+                progress: new Progress<IUploadProgress>(p =>
+                {
+                    if (_progress == null)
+                        return;
+
+                    switch (p.Status)
+                    {
+                        case UploadStatus.Starting:
+                        case UploadStatus.NotStarted:
+                            _progress.UploadProgress.ChangeState(UploadState.PendingUpload);
+                            break;
+                        case UploadStatus.Completed:
+                            _progress.UploadProgress.ChangeState(UploadState.Done);
+                            break;
+                        case UploadStatus.Uploading:
+                            _progress.UploadProgress.ChangeState(UploadState.Uploading);
+                            break;
+                    }
+
+                    _progress.UploadProgress.UploadedInBytes = p.BytesSent;
+                    _progress.OnUploadProgress();
+                }));
+        }
+
+        public async Task DownloadObjectAsync(string fileName, Stream stream)
+        {
+            await _client.DownloadObjectAsync(
```

Comment:
why awaiting?

Reply:
Removed & change to void.

Reply:
Don't you need to wait for the object to finish downloading?